generative/maisi/scripts/diff_model_infer.py

```diff
@@ -0,0 +1,222 @@
+            spacing_tensor = spacing_tensor * 1e2
+            print(f"spacing_tensor: {spacing_tensor}.")
+
+            top_region_index_tensor = top_region_index_tensor[np.newaxis, :]
+            bottom_region_index_tensor = bottom_region_index_tensor[np.newaxis, :]
+            spacing_tensor = spacing_tensor[np.newaxis, :]
+
+            top_region_index_tensor = torch.from_numpy(top_region_index_tensor).half().to(device)
+            bottom_region_index_tensor = torch.from_numpy(bottom_region_index_tensor).half().to(device)
+            spacing_tensor = torch.from_numpy(spacing_tensor).half().to(device)
+
+            image = noise
+
+            # synthesize latents
+            for t in tqdm(noise_scheduler.timesteps, ncols=110):
+                model_output = unet(
+                    x=image,
+                    timesteps=torch.Tensor((t,)).to(device),
+                    top_region_index_tensor=top_region_index_tensor,
+                    bottom_region_index_tensor=bottom_region_index_tensor,
+                    spacing_tensor=spacing_tensor,
+                )
+                image, _ = noise_scheduler.step(model_output, t, image)
+
+            synthetic_images = sliding_window_inference(
+                inputs=image,
+                roi_size=(
+                    min(output_size[0] // divisor // 4 * 3, 96),
+                    min(output_size[1] // divisor // 4 * 3, 96),
+                    min(output_size[2] // divisor // 4 * 3, 96),
+                ),
+                sw_batch_size=1,
+                predictor=recon_model,
+                mode="gaussian",
+                overlap=2.0 / 3.0,
+                sw_device=device,
+                device=device,
+            )
+
+            data = synthetic_images.squeeze().cpu().detach().numpy()
+            data = (data - b_min) / (b_max - b_min) * (a_max - a_min) + a_min
+            data = np.clip(data, a_min, a_max)
+            data = np.int16(data)
+
+            out_affine = np.eye(4)
+            for _k in range(3):
+                out_affine[_k, _k] = out_spacing[_k]
+            new_image = nib.Nifti1Image(data, affine=out_affine)
+
+        timestamp = datetime.now().strftime("%Y%m%d%H%M%S")
+        output_path = f"{args.output_dir}/{output_prefix}_seed{random_seed}_size{output_size[0]:d}x{output_size[1]:d}x{output_size[2]:d}_spacing{out_spacing[0]:.2f}x{out_spacing[1]:.2f}x{out_spacing[2]:.2f}_{timestamp}.nii.gz"
+        os.makedirs(os.path.dirname(output_path), exist_ok=True)
+        nib.save(new_image, output_path)
+        print(f"Saved {output_path}.")
+
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser(description="Diffusion Model Inference")
+    parser.add_argument("--env_config", type=str, required=True, help="Path to environment configuration file")
+    parser.add_argument("--model_config", type=str, required=True, help="Path to model configuration file")
```

Comment:
Same as above

Reply:
Hi @mingxin-zheng, I have updated the script. Please proceed. Thanks.